test/models/flava/test_flava.py

```diff
@@ -155,50 +188,84 @@ def setUp(self):
             use_image_masking=True,
         )
 
-        mm_encoder = nn.Identity()
-        image_to_mm_projection = nn.Identity()
-        text_to_mm_projection = nn.Identity()
-        self.flava = FLAVAModel(
-            image_encoder=self.image_encoder,
-            text_encoder=self.text_encoder,
+    @pytest.fixture
+    def mm_encoder(self):
+        return nn.Identity()
+
+    @pytest.fixture
+    def image_to_mm_projection(self):
+        return nn.Identity()
+
+    @pytest.fixture
+    def text_to_mm_projection(self):
+        return nn.Identity()
+
+    @pytest.fixture
+    def text_projection(self):
+        return nn.Identity()
+
+    @pytest.fixture
+    def image_projection(self):
+        return nn.Identity()
+
+    @pytest.fixture
+    def flava(
+        self,
+        image_encoder,
+        text_encoder,
+        mm_encoder,
+        image_to_mm_projection,
+        text_to_mm_projection,
+        text_projection,
+        image_projection,
+    ):
+        return FLAVAModel(
+            image_encoder=image_encoder,
+            text_encoder=text_encoder,
             mm_encoder=mm_encoder,
             image_to_mm_projection=image_to_mm_projection,
             text_to_mm_projection=text_to_mm_projection,
-            text_projection=nn.Identity(),
-            image_projection=nn.Identity(),
+            text_projection=text_projection,
+            image_projection=image_projection,
         )
 
-    def test_forward_image_text(self):
-        image = torch.ones(2, 3, 2, 2)
+    @pytest.fixture
+    def inputs(self):
+        image = torch.zeros(2, 3, 2, 2)
+        masked_image = torch.ones(2, 1)
         text = torch.ones(2, 3, dtype=torch.int32)
-        actual = self.flava(image, text)
+        masked_text = torch.ones(2, 3, dtype=torch.int32)
+        return image, masked_image, text, masked_text
+
+    def test_forward_image_text(self, image_encoder, text_encoder, flava, inputs):
+        image, _, text, _ = inputs
+        actual = flava(image, text)
         expected = FLAVAOutput(
             text_masked=TransformerOutput(),
             multimodal_masked=TransformerOutput(),
             multimodal=TransformerOutput(),
-            text=self.text_encoder(
+            text=text_encoder(
                 text, return_attn_weights=True, return_hidden_states=True
             ),
-            image=self.image_encoder(image),
-            image_masked=self.image_encoder(image),
+            image=image_encoder(image),
+            image_masked=image_encoder(image),
             projected_image_embeddings=actual.image.last_hidden_state[:, 0, :],
```

Comment:
`expected` depends on `actual`? !

```
projected_image_embeddings=actual.image.last_hidden_state[:, 0, :],
projected_text_embeddings=actual.text.last_hidden_state[:, 0, :],
```

Generally, it'd be good not to use components we test to calculate expected even if the values cannot be computed by hand to ensure isolation. For example, `text`, `image` and `image_masked` fields are computed from the low level components which are part of flava model. 

I'd suggest at least remove the explicit dependency on `actual` for the last two fields (and check if there're similar cases elsewhere). Your call for those that are derived from low level components.

Reply:
Yeah this is a good point. Let me derive these separately without relying on `actual`